core/src/main/scala/filodb.core/reprojector/FiloMemTable.scala

```diff
@@ -0,0 +1,216 @@
+  private var flushTask: Option[ScheduledFuture[_]] = None
+
+  private def getKeyMap(partition: PartitionKey): KeyMap = {
+    partKeyMap.getOrElse(partition, {
+      // Only hit this if partition isn't defined, so save expensive synchronization for initial cases only
+      partKeyMap.synchronized {
+        partKeyMap.getOrElseUpdate(partition, new KeyMap(projection.helper.ordering))
+      }
+    })
+  }
+
+  private def chunkRowIdToLong(chunkIndex: Int, rowNo: Int): Long =
+    (chunkIndex.toLong << 32) + rowNo
+
+  private def invokeAndCleanupCallbacks(rowsToAdd: Int): Unit = {
+    // Both this method and ingestRows modifies tempRows and callbacks
+    tempRows.synchronized {
+      // invoke callbacks
+      while (callbacks.nonEmpty && callbacks.head._1 < rowsToAdd && callbacks.head._2 < rowsToAdd) {
+        callbacks.head._3()
+        callbacks.remove(0, 1)
+      }
+
+      // Remove rows added from tempRows
+      tempRows.remove(0, rowsToAdd)
+
+      // Adjust remaining callback row indices
+      callbacks.zipWithIndex.foreach { case ((firstIdx, lastIdx, fn), i) =>
+        callbacks(i) = (Math.min(0, firstIdx - rowsToAdd), Math.min(0, lastIdx - rowsToAdd), fn)
+      }
+    }
+  }
+
+  // Converts rows to chunks, merging with previous chunks as needed, and invokes callbacks
+  // Also updates the partKeyMap which maps every sort key to its chunk ID and row index
+  private def flushRowsToChunks(): Unit = synchronized {
+    // VERY IMPORTANT: pass false to cancel so this task is not interrupted
+    // Cancel any other scheduled tasks, if one was scheduled
+    flushTask.foreach(_.cancel(false))
+    builder.reset()
+
+    // Last chunk written partial?  Get chunks out, populate builder, remove from chunks
+    if (readers.nonEmpty && readers.last.parsers(0).length < chunkSize) {
+      val reader = readers.last
+      for { i <- 0 until reader.parsers(0).length optimized } {
+        reader.rowNo = i
+        builder.addRow(reader)
+      }
+      chunks.remove(chunks.length - 1, 1)
+      readers.remove(readers.length - 1, 1)
+    }
+
+    // Add new rows to builder, adding to partKeyMap in the meantime
+    val baseLength = builder.builders.head.length   // Nonzero only if partial chunks added
+    val rowsToAdd = Math.min(chunkSize - baseLength, tempRows.length)
+    val nextChunkIndex = chunks.length
+    val sortKeyFunc = projection.sortKeyFunc
+
+    for { i <- 0 until rowsToAdd optimized } {
+      val row = tempRows(i)
+      builder.addRow(row)
+      val keyMap = getKeyMap(projection.partitionFunc(row))
+      keyMap.put(sortKeyFunc(row), chunkRowIdToLong(nextChunkIndex, baseLength + i))
+    }
+
+    // Add chunks
+    val chunkAray = new Array[ByteBuffer](filoSchema.length)
+    builder.convertToBytes().foreach { case (colName, bb) => chunkAray(colIdToNumber(colName)) = bb }
+    chunks += chunkAray
+    readers += new FastFiloRowReader(chunkAray, clazzes)
```

Comment:
Well, convertToBytes yields a map of column name to ByteBuffer, and we need the `colIdToNumber` mapping to convert the map back to an ordered array.  I suppose we could use for {} instead of foreach, or possibly even better is to have an ordered list of column names, then we could directly convert to an array.